File: python/examples/etl_example.py
from pyspark.sql import SparkSession, types as T, functions as F
import argparse

from dqsuite import DQSuiteContextBuilder

spark = SparkSession.builder.getOrCreate()

# Parse Arguments
parser = argparse.ArgumentParser()
parser.add_argument("--input_file_path", help="Path to source file")
parser.add_argument("--config_path", help="Path to config file")

args, _ = parser.parse_known_args()


input_file_path = args.input_file_path
config_path = args.config_path
dq_output_path = "../out"
dq_profile = False
dq_anomaly_detection = True

# Set up Data Quality Context
dqsContext = (
    DQSuiteContextBuilder.builder(spark)
    .withConfigPath(config_path)
    .withResultPath(f"{dq_output_path}/results")
    .withMetricsPath(f"{dq_output_path}/metrics")
    .build()
)

# Load data
df_raw = (
    spark.read
    .option("header", "true")
    .option("inferSchema", "false")
    .csv(input_file_path)
)

# Raw Data Quality Checks
dsContext = dqsContext.withDataset("sales_raw")

schemaCheckResult = dsContext.checkSchema(df_raw)
# TODO: save invalid rows
assert schemaCheckResult.isValid, f"Schema Check Failed: {schemaCheckResult}"
df_raw_valid = schemaCheckResult.validRows  # there are no invalid rows, otherwise the check would fail

if dq_profile:
    profilingResult = dsContext.profile(df_raw_valid)
    print(f"Profiling finished. Used {profilingResult.numRecordsUsedForProfiling()} for profiling")

validationResult = dsContext.validate(df_raw_valid, anomalyDetection=dq_anomaly_detection)
assert validationResult.status != "Error", f"Data Quality Checks Failed: {validationResult.checkResults}"
print(f"Data Quality Checks Passed: {validationResult.status}")

df = dsContext.postprocess(df_raw_valid)

# Data Transformation
df = (
    df
    .filter(F.col("invoice_item_number").isNotNull())
    .dropDuplicates(["invoice_item_number"])
    .withColumn("zip_code", F.regexp_extract("zip_code", r"(\d{5})", 1))
    .withColumn("year", F.lpad(F.year("date"), 4, "0"))
    .withColumn("month", F.lpad(F.month("date"), 2, "0"))
    .withColumn("day", F.lpad(F.dayofmonth("date"), 2, "0"))
    .sortWithinPartitions("year", "month", "day")
)

# Output Data Quality Check
dsContext = dqsContext.withDataset("sales_processed")

schemaCheckResult = dsContext.checkSchema(df)
assert schemaCheckResult.isValid, f"Schema Check Failed: {schemaCheckResult.missingColumns}"
df = schemaCheckResult.validRows

validationResult = dsContext.validate(df, anomalyDetection=dq_anomaly_detection)
assert validationResult.status != "Error", f"Data Quality Checks Failed: {validationResult.checkResults}"
print(f"Data Quality Checks Passed: {validationResult.status}")

File: examples/src/main/scala/ETLExample.scala
import com.amazon.deequ.checks.CheckStatus
import com.amazonaws.services.glue.{DynamicFrame, GlueContext}
import com.amazonaws.services.glue.log.GlueLogger
import com.amazonaws.services.glue.util.{GlueArgParser, Job}
import dqsuite.DQSuiteContextBuilder
import org.apache.spark.SparkContext
import org.apache.spark.sql.{functions => F, types => T}

import scala.jdk.CollectionConverters.mapAsJavaMapConverter

object ETLExample {
  private val sparkContext: SparkContext = new SparkContext()
  private val glueContext: GlueContext   = new GlueContext(sparkContext)
  private val spark                      = glueContext.getSparkSession
  private val logger                     = new GlueLogger()

  def main(sysArgs: Array[String]): Unit = {
    // * config_path: Path to the data quality configuration file. Can be a local file or a S3 URI.
    // * input_file_path: Path to the input file. Can be a local file or a S3 URI.
    val args = GlueArgParser.getResolvedOptions(sysArgs, Seq("JOB_NAME", "input_file_path", "config_path").toArray)
    Job.init(args("JOB_NAME"), glueContext, args.asJava)

    // Configure DQSuite
    val dqsContext = DQSuiteContextBuilder.builder
      .withConfigPath(args("config_path"))
      .withResultPath("./out/results")
      .withMetricsPath("./out/metrics")
      .withSparkSession(spark)
      .build

    // Load data
    val dfRaw = spark.read
      .option("header", "true")
      .option("inferSchema", "false")
      .csv(args("input_file_path"))

    // Run schema check
    var dsContext         = dqsContext.withDataset("sales_raw")
    var schemaCheckResult = dsContext.checkSchema(dfRaw)
    logger.info(
      s"Schema check finished. Found ${schemaCheckResult.numInvalidRows} invalid rows" +
        s" and ${schemaCheckResult.numValidRows} valid rows"
    )
    if (schemaCheckResult.missingColumns.nonEmpty) {
      logger.warn(s"Missing columns: ${schemaCheckResult.missingColumns.mkString(", ")}")
    }
    if (!schemaCheckResult.isValid) {
      logger.warn(s"Invalid rows: ${schemaCheckResult.invalidRows.show(50)}")
    }
    assert(schemaCheckResult.isValid)

    // Use valid rows
    val df_raw_valid = schemaCheckResult.validRows

    // Run profiler and validator
//    val profilingResult = dsContext.profile(df_raw_valid)
//    logger.info(s"Profiling finished. Used ${profilingResult.numRecordsUsedForProfiling} for profiling")

    var validationResult = dsContext.validate(df_raw_valid)
    validationResult.status match {
      case CheckStatus.Success => logger.info("Validation succeeded")
      case CheckStatus.Warning => logger.warn("Validation succeeded with warnings")
      case CheckStatus.Error   => throw new RuntimeException("Validation failed")
    }

    var df = dsContext.postprocess(df_raw_valid)

    // Transform data
    df = (
      df.filter(F.col("invoice_item_number").isNotNull)
        .dropDuplicates(Seq("invoice_item_number"))
        .withColumn("zip_code", F.regexp_extract(F.col("zip_code"), "[0-9]{5}", 0))
        .withColumn("year", F.lpad(F.year(F.col("date")), 4, "0"))
        .withColumn("month", F.lpad(F.month(F.col("date")), 2, "0"))
        .withColumn("day", F.lpad(F.dayofmonth(F.col("date")), 2, "0"))
        .sortWithinPartitions(
          "year",
          "month",
          "day"
        )
    )

    // Run dataquality
    dsContext = dqsContext.withDataset("sales_processed")

    schemaCheckResult = dsContext.checkSchema(df)
    logger.info(
      s"Schema check finished. Found ${schemaCheckResult.numInvalidRows} invalid rows" +
        s" and ${schemaCheckResult.numValidRows} valid rows"
    )
    if (schemaCheckResult.missingColumns.nonEmpty) {
      logger.warn(s"Missing columns: ${schemaCheckResult.missingColumns.mkString(", ")}")
    }
    if (!schemaCheckResult.isValid) {
      logger.warn(s"Invalid rows: ${schemaCheckResult.invalidRows.show(50)}")
    }
    assert(schemaCheckResult.isValid)

    validationResult = dsContext.validate(df)
    validationResult.status match {
      case CheckStatus.Success => logger.info("Validation succeeded")
      case CheckStatus.Warning => logger.warn("Validation succeeded with warnings")
      case CheckStatus.Error   => throw new RuntimeException("Validation failed")
    }

  }
}
